<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1">

<style type="text/css">
@font-face {
font-family: octicons-link;
src: url(data:font/woff;charset=utf-8;base64,d09GRgABAAAAAAZwABAAAAAACFQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAABEU0lHAAAGaAAAAAgAAAAIAAAAAUdTVUIAAAZcAAAACgAAAAoAAQAAT1MvMgAAAyQAAABJAAAAYFYEU3RjbWFwAAADcAAAAEUAAACAAJThvmN2dCAAAATkAAAABAAAAAQAAAAAZnBnbQAAA7gAAACyAAABCUM+8IhnYXNwAAAGTAAAABAAAAAQABoAI2dseWYAAAFsAAABPAAAAZwcEq9taGVhZAAAAsgAAAA0AAAANgh4a91oaGVhAAADCAAAABoAAAAkCA8DRGhtdHgAAAL8AAAADAAAAAwGAACfbG9jYQAAAsAAAAAIAAAACABiATBtYXhwAAACqAAAABgAAAAgAA8ASm5hbWUAAAToAAABQgAAAlXu73sOcG9zdAAABiwAAAAeAAAAME3QpOBwcmVwAAAEbAAAAHYAAAB/aFGpk3jaTY6xa8JAGMW/O62BDi0tJLYQincXEypYIiGJjSgHniQ6umTsUEyLm5BV6NDBP8Tpts6F0v+k/0an2i+itHDw3v2+9+DBKTzsJNnWJNTgHEy4BgG3EMI9DCEDOGEXzDADU5hBKMIgNPZqoD3SilVaXZCER3/I7AtxEJLtzzuZfI+VVkprxTlXShWKb3TBecG11rwoNlmmn1P2WYcJczl32etSpKnziC7lQyWe1smVPy/Lt7Kc+0vWY/gAgIIEqAN9we0pwKXreiMasxvabDQMM4riO+qxM2ogwDGOZTXxwxDiycQIcoYFBLj5K3EIaSctAq2kTYiw+ymhce7vwM9jSqO8JyVd5RH9gyTt2+J/yUmYlIR0s04n6+7Vm1ozezUeLEaUjhaDSuXHwVRgvLJn1tQ7xiuVv/ocTRF42mNgZGBgYGbwZOBiAAFGJBIMAAizAFoAAABiAGIAznjaY2BkYGAA4in8zwXi+W2+MjCzMIDApSwvXzC97Z4Ig8N/BxYGZgcgl52BCSQKAA3jCV8CAABfAAAAAAQAAEB42mNgZGBg4f3vACQZQABIMjKgAmYAKEgBXgAAeNpjYGY6wTiBgZWBg2kmUxoDA4MPhGZMYzBi1AHygVLYQUCaawqDA4PChxhmh/8ODDEsvAwHgMKMIDnGL0x7gJQCAwMAJd4MFwAAAHjaY2BgYGaA4DAGRgYQkAHyGMF8NgYrIM3JIAGVYYDT+AEjAwuDFpBmA9KMDEwMCh9i/v8H8sH0/4dQc1iAmAkALaUKLgAAAHjaTY9LDsIgEIbtgqHUPpDi3gPoBVyRTmTddOmqTXThEXqrob2gQ1FjwpDvfwCBdmdXC5AVKFu3e5MfNFJ29KTQT48Ob9/lqYwOGZxeUelN2U2R6+cArgtCJpauW7UQBqnFkUsjAY/kOU1cP+DAgvxwn1chZDwUbd6CFimGXwzwF6tPbFIcjEl+vvmM/byA48e6tWrKArm4ZJlCbdsrxksL1AwWn/yBSJKpYbq8AXaaTb8AAHja28jAwOC00ZrBeQNDQOWO//sdBBgYGRiYWYAEELEwMTE4uzo5Zzo5b2BxdnFOcALxNjA6b2ByTswC8jYwg0VlNuoCTWAMqNzMzsoK1rEhNqByEyerg5PMJlYuVueETKcd/89uBpnpvIEVomeHLoMsAAe1Id4AAAAAAAB42oWQT07CQBTGv0JBhagk7HQzKxca2sJCE1hDt4QF+9JOS0nbaaYDCQfwCJ7Au3AHj+LO13FMmm6cl7785vven0kBjHCBhfpYuNa5Ph1c0e2Xu3jEvWG7UdPDLZ4N92nOm+EBXuAbHmIMSRMs+4aUEd4Nd3CHD8NdvOLTsA2GL8M9PODbcL+hD7C1xoaHeLJSEao0FEW14ckxC+TU8TxvsY6X0eLPmRhry2WVioLpkrbp84LLQPGI7c6sOiUzpWIWS5GzlSgUzzLBSikOPFTOXqly7rqx0Z1Q5BAIoZBSFihQYQOOBEdkCOgXTOHA07HAGjGWiIjaPZNW13/+lm6S9FT7rLHFJ6fQbkATOG1j2OFMucKJJsxIVfQORl+9Jyda6Sl1dUYhSCm1dyClfoeDve4qMYdLEbfqHf3O/AdDumsjAAB42mNgYoAAZQYjBmyAGYQZmdhL8zLdDEydARfoAqIAAAABAAMABwAKABMAB///AA8AAQAAAAAAAAAAAAAAAAABAAAAAA==) format('woff');
}
body {
-webkit-text-size-adjust: 100%;
text-size-adjust: 100%;
color: #333;
font-family: "Helvetica Neue", Helvetica, "Segoe UI", Arial, freesans, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
font-size: 16px;
line-height: 1.6;
word-wrap: break-word;
}
a {
background-color: transparent;
}
a:active,
a:hover {
outline: 0;
}
strong {
font-weight: bold;
}
h1 {
font-size: 2em;
margin: 0.67em 0;
}
img {
border: 0;
}
hr {
box-sizing: content-box;
height: 0;
}
pre {
overflow: auto;
}
code,
kbd,
pre {
font-family: monospace, monospace;
font-size: 1em;
}
input {
color: inherit;
font: inherit;
margin: 0;
}
html input[disabled] {
cursor: default;
}
input {
line-height: normal;
}
input[type="checkbox"] {
box-sizing: border-box;
padding: 0;
}
table {
border-collapse: collapse;
border-spacing: 0;
}
td,
th {
padding: 0;
}
* {
box-sizing: border-box;
}
input {
font: 13px / 1.4 Helvetica, arial, nimbussansl, liberationsans, freesans, clean, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol";
}
a {
color: #4078c0;
text-decoration: none;
}
a:hover,
a:active {
text-decoration: underline;
}
hr {
height: 0;
margin: 15px 0;
overflow: hidden;
background: transparent;
border: 0;
border-bottom: 1px solid #ddd;
}
hr:before {
display: table;
content: "";
}
hr:after {
display: table;
clear: both;
content: "";
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 15px;
margin-bottom: 15px;
line-height: 1.1;
}
h1 {
font-size: 30px;
}
h2 {
font-size: 21px;
}
h3 {
font-size: 16px;
}
h4 {
font-size: 14px;
}
h5 {
font-size: 12px;
}
h6 {
font-size: 11px;
}
blockquote {
margin: 0;
}
ul,
ol {
padding: 0;
margin-top: 0;
margin-bottom: 0;
}
ol ol,
ul ol {
list-style-type: lower-roman;
}
ul ul ol,
ul ol ol,
ol ul ol,
ol ol ol {
list-style-type: lower-alpha;
}
dd {
margin-left: 0;
}
code {
font-family: Consolas, "Liberation Mono", Menlo, Courier, monospace;
font-size: 12px;
}
pre {
margin-top: 0;
margin-bottom: 0;
font: 12px Consolas, "Liberation Mono", Menlo, Courier, monospace;
}
.select::-ms-expand {
opacity: 0;
}
.octicon {
font: normal normal normal 16px/1 octicons-link;
display: inline-block;
text-decoration: none;
text-rendering: auto;
-webkit-font-smoothing: antialiased;
-moz-osx-font-smoothing: grayscale;
-webkit-user-select: none;
-moz-user-select: none;
-ms-user-select: none;
user-select: none;
}
.octicon-link:before {
content: '\f05c';
}
.markdown-body:before {
display: table;
content: "";
}
.markdown-body:after {
display: table;
clear: both;
content: "";
}
.markdown-body>*:first-child {
margin-top: 0 !important;
}
.markdown-body>*:last-child {
margin-bottom: 0 !important;
}
a:not([href]) {
color: inherit;
text-decoration: none;
}
.anchor {
display: inline-block;
padding-right: 2px;
margin-left: -18px;
}
.anchor:focus {
outline: none;
}
h1,
h2,
h3,
h4,
h5,
h6 {
margin-top: 1em;
margin-bottom: 16px;
font-weight: bold;
line-height: 1.4;
}
h1 .octicon-link,
h2 .octicon-link,
h3 .octicon-link,
h4 .octicon-link,
h5 .octicon-link,
h6 .octicon-link {
color: #000;
vertical-align: middle;
visibility: hidden;
}
h1:hover .anchor,
h2:hover .anchor,
h3:hover .anchor,
h4:hover .anchor,
h5:hover .anchor,
h6:hover .anchor {
text-decoration: none;
}
h1:hover .anchor .octicon-link,
h2:hover .anchor .octicon-link,
h3:hover .anchor .octicon-link,
h4:hover .anchor .octicon-link,
h5:hover .anchor .octicon-link,
h6:hover .anchor .octicon-link {
visibility: visible;
}
h1 {
padding-bottom: 0.3em;
font-size: 2.25em;
line-height: 1.2;
border-bottom: 1px solid #eee;
}
h1 .anchor {
line-height: 1;
}
h2 {
padding-bottom: 0.3em;
font-size: 1.75em;
line-height: 1.225;
border-bottom: 1px solid #eee;
}
h2 .anchor {
line-height: 1;
}
h3 {
font-size: 1.5em;
line-height: 1.43;
}
h3 .anchor {
line-height: 1.2;
}
h4 {
font-size: 1.25em;
}
h4 .anchor {
line-height: 1.2;
}
h5 {
font-size: 1em;
}
h5 .anchor {
line-height: 1.1;
}
h6 {
font-size: 1em;
color: #777;
}
h6 .anchor {
line-height: 1.1;
}
p,
blockquote,
ul,
ol,
dl,
table,
pre {
margin-top: 0;
margin-bottom: 16px;
}
hr {
height: 4px;
padding: 0;
margin: 16px 0;
background-color: #e7e7e7;
border: 0 none;
}
ul,
ol {
padding-left: 2em;
}
ul ul,
ul ol,
ol ol,
ol ul {
margin-top: 0;
margin-bottom: 0;
}
li>p {
margin-top: 16px;
}
dl {
padding: 0;
}
dl dt {
padding: 0;
margin-top: 16px;
font-size: 1em;
font-style: italic;
font-weight: bold;
}
dl dd {
padding: 0 16px;
margin-bottom: 16px;
}
blockquote {
padding: 0 15px;
color: #777;
border-left: 4px solid #ddd;
}
blockquote>:first-child {
margin-top: 0;
}
blockquote>:last-child {
margin-bottom: 0;
}
table {
display: block;
width: 100%;
overflow: auto;
word-break: normal;
word-break: keep-all;
}
table th {
font-weight: bold;
}
table th,
table td {
padding: 6px 13px;
border: 1px solid #ddd;
}
table tr {
background-color: #fff;
border-top: 1px solid #ccc;
}
table tr:nth-child(2n) {
background-color: #f8f8f8;
}
img {
max-width: 100%;
box-sizing: content-box;
background-color: #fff;
}
code {
padding: 0;
padding-top: 0.2em;
padding-bottom: 0.2em;
margin: 0;
font-size: 85%;
background-color: rgba(0,0,0,0.04);
border-radius: 3px;
}
code:before,
code:after {
letter-spacing: -0.2em;
content: "\00a0";
}
pre>code {
padding: 0;
margin: 0;
font-size: 100%;
word-break: normal;
white-space: pre;
background: transparent;
border: 0;
}
.highlight {
margin-bottom: 16px;
}
.highlight pre,
pre {
padding: 16px;
overflow: auto;
font-size: 85%;
line-height: 1.45;
background-color: #f7f7f7;
border-radius: 3px;
}
.highlight pre {
margin-bottom: 0;
word-break: normal;
}
pre {
word-wrap: normal;
}
pre code {
display: inline;
max-width: initial;
padding: 0;
margin: 0;
overflow: initial;
line-height: inherit;
word-wrap: normal;
background-color: transparent;
border: 0;
}
pre code:before,
pre code:after {
content: normal;
}
kbd {
display: inline-block;
padding: 3px 5px;
font-size: 11px;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.pl-c {
color: #969896;
}
.pl-c1,
.pl-s .pl-v {
color: #0086b3;
}
.pl-e,
.pl-en {
color: #795da3;
}
.pl-s .pl-s1,
.pl-smi {
color: #333;
}
.pl-ent {
color: #63a35c;
}
.pl-k {
color: #a71d5d;
}
.pl-pds,
.pl-s,
.pl-s .pl-pse .pl-s1,
.pl-sr,
.pl-sr .pl-cce,
.pl-sr .pl-sra,
.pl-sr .pl-sre {
color: #183691;
}
.pl-v {
color: #ed6a43;
}
.pl-id {
color: #b52a1d;
}
.pl-ii {
background-color: #b52a1d;
color: #f8f8f8;
}
.pl-sr .pl-cce {
color: #63a35c;
font-weight: bold;
}
.pl-ml {
color: #693a17;
}
.pl-mh,
.pl-mh .pl-en,
.pl-ms {
color: #1d3e81;
font-weight: bold;
}
.pl-mq {
color: #008080;
}
.pl-mi {
color: #333;
font-style: italic;
}
.pl-mb {
color: #333;
font-weight: bold;
}
.pl-md {
background-color: #ffecec;
color: #bd2c00;
}
.pl-mi1 {
background-color: #eaffea;
color: #55a532;
}
.pl-mdr {
color: #795da3;
font-weight: bold;
}
.pl-mo {
color: #1d3e81;
}
kbd {
display: inline-block;
padding: 3px 5px;
font: 11px Consolas, "Liberation Mono", Menlo, Courier, monospace;
line-height: 10px;
color: #555;
vertical-align: middle;
background-color: #fcfcfc;
border: solid 1px #ccc;
border-bottom-color: #bbb;
border-radius: 3px;
box-shadow: inset 0 -1px 0 #bbb;
}
.task-list-item {
list-style-type: none;
}
.task-list-item+.task-list-item {
margin-top: 3px;
}
.task-list-item input {
margin: 0 0.35em 0.25em -1.6em;
vertical-align: middle;
}
:checked+.radio-label {
z-index: 1;
position: relative;
border-color: #4078c0;
}
.sourceLine {
display: inline-block;
}
code .kw { color: #000000; }
code .dt { color: #ed6a43; }
code .dv { color: #009999; }
code .bn { color: #009999; }
code .fl { color: #009999; }
code .ch { color: #009999; }
code .st { color: #183691; }
code .co { color: #969896; }
code .ot { color: #0086b3; }
code .al { color: #a61717; }
code .fu { color: #63a35c; }
code .er { color: #a61717; background-color: #e3d2d2; }
code .wa { color: #000000; }
code .cn { color: #008080; }
code .sc { color: #008080; }
code .vs { color: #183691; }
code .ss { color: #183691; }
code .im { color: #000000; }
code .va {color: #008080; }
code .cf { color: #000000; }
code .op { color: #000000; }
code .bu { color: #000000; }
code .ex { color: #000000; }
code .pp { color: #999999; }
code .at { color: #008080; }
code .do { color: #969896; }
code .an { color: #008080; }
code .cv { color: #008080; }
code .in { color: #008080; }
</style>
<style>
body {
  box-sizing: border-box;
  min-width: 200px;
  max-width: 980px;
  margin: 0 auto;
  padding: 45px;
  padding-top: 0px;
}
</style>

</head>

<body>

<h1 id="nearest-centroid-classification-with-bigquery">Nearest Centroid Classification with BigQuery</h1>
<h1 id="isb-cgc-community-notebooks">ISB-CGC Community Notebooks</h1>
<p>Check out more notebooks at our <a href="https://github.com/isb-cgc/Community-Notebooks">Community Notebooks Repository</a>!</p>
<pre><code>Title:   How to Use BigQuery to Preform Nearest Centroid Classification
Author:  Lauren Hagen
Created: 2019-12-17
URL:     https://github.com/isb-cgc/Community-Notebooks/blob/master/Notebooks/Nearest_Centroid_Classification_with_BigQuery.ipynb
Purpose: Demostrate using BigQuery to categorize patients based on gene expression using the Nearest Centroid Classification.
Notes: 
</code></pre>
<h1 id="introduction">Introduction</h1>
<h2 id="overview">Overview</h2>
<p>This notebook is to demonstrate how to use BigQuery to categorize patients based on gene expression using the Nearest Centroid Classifier. We will be using the Kidney renal papillary cell carcinoma (KIRP) study from the The Cancer Genome Atlas (TCGA) as an example data set and will use the RNA Sequence Gene Expression Data.</p>
<h2 id="what-is-nearest-centroid-classifier">What is Nearest Centroid Classifier?</h2>
<p>Nearest Centroid Classifier assigns a label based on the nearest mean (centroid) of the training samples that the observation is closest to<sup>1</sup>. This classifier is an example of supervised learning and does not create a model for use later<sup>2</sup>.</p>
<p>Before we get started we will need to load the BigQuery module, authenticate ourselves, create a client variable, and load necessary libraries.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(bigrquery)</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="kw">library</span>(stringr)</a>
<a class="sourceLine" id="cb2-3" title="3"><span class="kw">library</span>(caret)</a></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" title="1">billing &lt;-<span class="st"> &#39;isb-cgc-02-0001&#39;</span> <span class="co"># Insert your project ID in the &#39;&#39;</span></a>
<a class="sourceLine" id="cb3-2" title="2"><span class="cf">if</span> (billing <span class="op">==</span><span class="st"> &#39;your_project_number&#39;</span>) {</a>
<a class="sourceLine" id="cb3-3" title="3">  <span class="kw">print</span>(<span class="st">&#39;Please update the project number with your Google Cloud Project&#39;</span>)</a>
<a class="sourceLine" id="cb3-4" title="4">}</a></code></pre></div>
<h1 id="select-genes">Select Genes</h1>
<p>We will be using the top three genes with the highest mean expression as the genes we will use to categorize the clinical stage for each case.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" title="1">top_<span class="dv">20</span>_highest_mean_gene_exp &lt;-<span class="st"> &quot;SELECT gene_name, AVG(HTSeq__FPKM_UQ) as m</span></a>
<a class="sourceLine" id="cb4-2" title="2"><span class="st">FROM `isb-cgc.TCGA_hg38_data_v0.RNAseq_Gene_Expression`</span></a>
<a class="sourceLine" id="cb4-3" title="3"><span class="st">WHERE project_short_name = &#39;TCGA-KIRP&#39;</span></a>
<a class="sourceLine" id="cb4-4" title="4"><span class="st">GROUP BY gene_name</span></a>
<a class="sourceLine" id="cb4-5" title="5"><span class="st">ORDER BY m DESC</span></a>
<a class="sourceLine" id="cb4-6" title="6"><span class="st">LIMIT 20&quot;</span></a></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1"><span class="co"># To see the R console output with query processing information, turn queit to FALSE</span></a>
<a class="sourceLine" id="cb5-2" title="2">top_genes_result &lt;-<span class="st"> </span><span class="kw">bq_project_query</span>(billing, top_<span class="dv">20</span>_highest_mean_gene_exp, <span class="dt">quiet =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-3" title="3"><span class="co"># Transform the query result into a tibble</span></a>
<a class="sourceLine" id="cb5-4" title="4">top_genes_result &lt;-<span class="st"> </span><span class="kw">bq_table_download</span>(top_genes_result, <span class="dt">quiet =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-5" title="5">top_genes_result</a></code></pre></div>
<pre><code>## # A tibble: 20 x 2
##    gene_name          m
##    &lt;chr&gt;          &lt;dbl&gt;
##  1 MT-CO3    527567678.
##  2 MT-CO1    463810408.
##  3 MT-CO2    437422084.
##  4 MT-ND4    418890462.
##  5 MT-RNR2   338365522.
##  6 MT-ATP6   271538283.
##  7 MT-ND3    247704187.
##  8 MT-CYB    239706855.
##  9 FTL       163189324.
## 10 MT-ND2    162676400.
## 11 MT-ND1    160206384.
## 12 MT-ND4L   129925423.
## 13 MT-ATP8   106401853.
## 14 MT-ND6    101825708.
## 15 MT-ND5     84101367.
## 16 MT-RNR1    82018149.
## 17 TMSB10     72683143.
## 18 MT-TP      68068246.
## 19 SPP1       63083630.
## 20 MTATP6P1   40230222.
</code></pre>
<h1 id="nearest-centroids-in-bigquery">Nearest Centroids in BigQuery</h1>
<h2 id="what-we-are-going-to-use-nearest-centroids-for">What we are going to use Nearest Centroids for?</h2>
<p>We will be attempting to predict the clinical stage for a case based on a certain subset of genes and their expression levels.</p>
<p>First we will filter the TCGA clinical table for our target disease (KIRP) and not include the cases where the clinical stage is missing. We need there to be no missing clinical stage features due to the nature of k nearest algorithms.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1">filtered_clin &lt;-<span class="st"> &quot;</span></a>
<a class="sourceLine" id="cb7-2" title="2"><span class="st">WITH</span></a>
<a class="sourceLine" id="cb7-3" title="3"><span class="st">  clin AS (</span></a>
<a class="sourceLine" id="cb7-4" title="4"><span class="st">  SELECT</span></a>
<a class="sourceLine" id="cb7-5" title="5"><span class="st">    case_barcode,</span></a>
<a class="sourceLine" id="cb7-6" title="6"><span class="st">    clinical_stage    </span></a>
<a class="sourceLine" id="cb7-7" title="7"><span class="st">  FROM</span></a>
<a class="sourceLine" id="cb7-8" title="8"><span class="st">    `isb-cgc.TCGA_bioclin_v0.Clinical`</span></a>
<a class="sourceLine" id="cb7-9" title="9"><span class="st">  WHERE</span></a>
<a class="sourceLine" id="cb7-10" title="10"><span class="st">    disease_code = &#39;KIRP&#39;</span></a>
<a class="sourceLine" id="cb7-11" title="11"><span class="st">    AND clinical_stage &lt;&gt; &#39;NULL&#39;</span></a>
<a class="sourceLine" id="cb7-12" title="12"><span class="st">  ), &quot;</span></a></code></pre></div>
<p>Next we will want to get the gene expressions for the genes that we are going to use to attempt to identify the clinical stage. We will filter for the 3 genes that we want, then we will label each case with whether it will be in the training or testing group. For more information on randomization in BigQuery, please see the <a href="https://github.com/isb-cgc/Community-Notebooks/blob/master/Notebooks/How_to_create_a_random_sample_in_bigquery.ipynb">How to Create a Random Sample in BigQuery Notebook</a> in the <a href="https://github.com/isb-cgc/Community-Notebooks">ISB-CGC Community Notebook Repository</a>.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" title="1">filtered_expr &lt;-<span class="st"> &quot;</span></a>
<a class="sourceLine" id="cb8-2" title="2"><span class="st">expr AS (</span></a>
<a class="sourceLine" id="cb8-3" title="3"><span class="st">  SELECT</span></a>
<a class="sourceLine" id="cb8-4" title="4"><span class="st">    case_barcode,</span></a>
<a class="sourceLine" id="cb8-5" title="5"><span class="st">    IF ( MOD(ABS(FARM_FINGERPRINT(case_barcode)),10) &gt; 1, &#39;TRAIN&#39;, &#39;TEST&#39;) as class,</span></a>
<a class="sourceLine" id="cb8-6" title="6"><span class="st">    gene_name,</span></a>
<a class="sourceLine" id="cb8-7" title="7"><span class="st">    HTSeq__FPKM_UQ</span></a>
<a class="sourceLine" id="cb8-8" title="8"><span class="st">  FROM</span></a>
<a class="sourceLine" id="cb8-9" title="9"><span class="st">    `isb-cgc.TCGA_hg38_data_v0.RNAseq_Gene_Expression`</span></a>
<a class="sourceLine" id="cb8-10" title="10"><span class="st">  WHERE</span></a>
<a class="sourceLine" id="cb8-11" title="11"><span class="st">    project_short_name = &#39;TCGA-KIRP&#39;</span></a>
<a class="sourceLine" id="cb8-12" title="12"><span class="st">    AND gene_name IN (&#39;MT-CO3&#39;,&#39;MT-CO1&#39;,&#39;MT-CO2&#39;)</span></a>
<a class="sourceLine" id="cb8-13" title="13"><span class="st">    AND (</span></a>
<a class="sourceLine" id="cb8-14" title="14"><span class="st">    case_barcode IN (select case_barcode from clin)</span></a>
<a class="sourceLine" id="cb8-15" title="15"><span class="st">    )</span></a>
<a class="sourceLine" id="cb8-16" title="16"><span class="st">  GROUP BY case_barcode, gene_name, HTSeq__FPKM_UQ</span></a>
<a class="sourceLine" id="cb8-17" title="17"><span class="st">  ),&quot;</span></a></code></pre></div>
<p>We will then find the centroids or means for each clinical stage within the training data.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1">centroids &lt;-<span class="st"> &quot;</span></a>
<a class="sourceLine" id="cb9-2" title="2"><span class="st"> mean AS (</span></a>
<a class="sourceLine" id="cb9-3" title="3"><span class="st">  SELECT</span></a>
<a class="sourceLine" id="cb9-4" title="4"><span class="st">    expr.class,</span></a>
<a class="sourceLine" id="cb9-5" title="5"><span class="st">    clin.clinical_stage,</span></a>
<a class="sourceLine" id="cb9-6" title="6"><span class="st">    AVG(CASE</span></a>
<a class="sourceLine" id="cb9-7" title="7"><span class="st">        WHEN gene_name = &#39;MT-CO3&#39; THEN HTSeq__FPKM_UQ</span></a>
<a class="sourceLine" id="cb9-8" title="8"><span class="st">    END</span></a>
<a class="sourceLine" id="cb9-9" title="9"><span class="st">      ) AS gene1,</span></a>
<a class="sourceLine" id="cb9-10" title="10"><span class="st">    AVG(CASE</span></a>
<a class="sourceLine" id="cb9-11" title="11"><span class="st">        WHEN gene_name = &#39;MT-CO1&#39; THEN HTSeq__FPKM_UQ</span></a>
<a class="sourceLine" id="cb9-12" title="12"><span class="st">    END</span></a>
<a class="sourceLine" id="cb9-13" title="13"><span class="st">      ) AS gene2,</span></a>
<a class="sourceLine" id="cb9-14" title="14"><span class="st">    AVG(CASE</span></a>
<a class="sourceLine" id="cb9-15" title="15"><span class="st">        WHEN gene_name = &#39;MT-CO2&#39; THEN HTSeq__FPKM_UQ</span></a>
<a class="sourceLine" id="cb9-16" title="16"><span class="st">    END</span></a>
<a class="sourceLine" id="cb9-17" title="17"><span class="st">      ) AS gene3</span></a>
<a class="sourceLine" id="cb9-18" title="18"><span class="st">  FROM</span></a>
<a class="sourceLine" id="cb9-19" title="19"><span class="st">    expr</span></a>
<a class="sourceLine" id="cb9-20" title="20"><span class="st">  JOIN</span></a>
<a class="sourceLine" id="cb9-21" title="21"><span class="st">    clin</span></a>
<a class="sourceLine" id="cb9-22" title="22"><span class="st">  ON</span></a>
<a class="sourceLine" id="cb9-23" title="23"><span class="st">    expr.case_barcode = clin.case_barcode</span></a>
<a class="sourceLine" id="cb9-24" title="24"><span class="st">  WHERE</span></a>
<a class="sourceLine" id="cb9-25" title="25"><span class="st">    expr.class = &#39;TRAIN&#39;</span></a>
<a class="sourceLine" id="cb9-26" title="26"><span class="st">  GROUP BY</span></a>
<a class="sourceLine" id="cb9-27" title="27"><span class="st">    expr.class,</span></a>
<a class="sourceLine" id="cb9-28" title="28"><span class="st">    clin.clinical_stage),&quot;</span></a></code></pre></div>
<p>We will also need the testing data to have the genes that we are going to use in the classifier in separate columns for the test cases.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" title="1">test &lt;-<span class="st"> &quot;</span></a>
<a class="sourceLine" id="cb10-2" title="2"><span class="st">test AS (</span></a>
<a class="sourceLine" id="cb10-3" title="3"><span class="st">  SELECT</span></a>
<a class="sourceLine" id="cb10-4" title="4"><span class="st">    case_barcode,</span></a>
<a class="sourceLine" id="cb10-5" title="5"><span class="st">    class,</span></a>
<a class="sourceLine" id="cb10-6" title="6"><span class="st">    MAX(CASE WHEN gene_name = &#39;MT-CO3&#39; THEN HTSeq__FPKM_UQ END) AS gene1,</span></a>
<a class="sourceLine" id="cb10-7" title="7"><span class="st">    MAX(CASE WHEN gene_name = &#39;MT-CO1&#39; THEN HTSeq__FPKM_UQ END) AS gene2,</span></a>
<a class="sourceLine" id="cb10-8" title="8"><span class="st">    MAX(CASE WHEN gene_name = &#39;MT-CO2&#39; THEN HTSeq__FPKM_UQ END) AS gene3</span></a>
<a class="sourceLine" id="cb10-9" title="9"><span class="st">  FROM</span></a>
<a class="sourceLine" id="cb10-10" title="10"><span class="st">    expr</span></a>
<a class="sourceLine" id="cb10-11" title="11"><span class="st">  WHERE</span></a>
<a class="sourceLine" id="cb10-12" title="12"><span class="st">    class = &#39;TEST&#39;</span></a>
<a class="sourceLine" id="cb10-13" title="13"><span class="st">  GROUP BY case_barcode, class),&quot;</span></a></code></pre></div>
<p>This next section of the query will find the euclidean distance for each case in the testing data set<sup>3</sup>. The euclidean distance can be found by the following equation<sup>4</sup>:</p>
<p>[d(q,p) = \sqrt{(q_1-p_1)^2+(q_2-p_2)^2+...+(q_n-p_n)^2}]</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1">dist &lt;-<span class="st"> &quot;</span></a>
<a class="sourceLine" id="cb11-2" title="2"><span class="st">distance AS (SELECT</span></a>
<a class="sourceLine" id="cb11-3" title="3"><span class="st">  case_barcode,</span></a>
<a class="sourceLine" id="cb11-4" title="4"><span class="st">  gene1,</span></a>
<a class="sourceLine" id="cb11-5" title="5"><span class="st">  gene2,</span></a>
<a class="sourceLine" id="cb11-6" title="6"><span class="st">  SQRT(POWER((SELECT gene1 FROM mean WHERE clinical_stage = &#39;Stage I&#39;)-gene1, 2) + POWER((SELECT gene2 FROM mean WHERE clinical_stage = &#39;Stage I&#39;)-gene2, 2) + POWER((SELECT gene3 FROM mean WHERE clinical_stage = &#39;Stage I&#39;)-gene3, 2)) as stage1,</span></a>
<a class="sourceLine" id="cb11-7" title="7"><span class="st">  SQRT(POWER((SELECT gene1 FROM mean WHERE clinical_stage = &#39;Stage II&#39;)- gene1, 2) + POWER((SELECT gene2 FROM mean WHERE clinical_stage = &#39;Stage II&#39;)-gene2, 2) + POWER((SELECT gene3 FROM mean WHERE clinical_stage = &#39;Stage II&#39;)-gene3, 2)) as stage2,</span></a>
<a class="sourceLine" id="cb11-8" title="8"><span class="st">  SQRT(POWER((SELECT gene1 FROM mean WHERE clinical_stage = &#39;Stage III&#39;)-gene1, 2) + POWER((SELECT gene2 FROM mean WHERE clinical_stage = &#39;Stage III&#39;)-gene2, 2) + POWER((SELECT gene3 FROM mean WHERE clinical_stage = &#39;Stage III&#39;)-gene3, 2)) as stage3,</span></a>
<a class="sourceLine" id="cb11-9" title="9"><span class="st">  SQRT(POWER((SELECT gene1 FROM mean WHERE clinical_stage = &#39;Stage IV&#39;)-gene1, 2) + POWER((SELECT gene2 FROM mean WHERE clinical_stage = &#39;Stage IV&#39;)-gene2, 2) + POWER((SELECT gene3 FROM mean WHERE clinical_stage = &#39;Stage IV&#39;)-gene3, 2)) as stage4,</span></a>
<a class="sourceLine" id="cb11-10" title="10"><span class="st">FROM test),</span></a>
<a class="sourceLine" id="cb11-11" title="11"><span class="st">&quot;</span></a></code></pre></div>
<p>Finally, we will calculate which category or stage has the smallest distance from the centroid and assign that stage to the test case. We will also include the case barcode and actual clinical stage as columns in our final result to assist with calculating the accuracy of the classification.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" title="1">pred &lt;-<span class="st"> &quot;</span></a>
<a class="sourceLine" id="cb12-2" title="2"><span class="st">pred AS (SELECT case_barcode,</span></a>
<a class="sourceLine" id="cb12-3" title="3"><span class="st">  (CASE WHEN stage1&lt;stage2 AND stage1&lt;stage3 AND stage1&lt;stage4 THEN &#39;Stage I&#39;</span></a>
<a class="sourceLine" id="cb12-4" title="4"><span class="st">  WHEN stage2&lt;stage1 AND stage2&lt;stage3 AND stage2&lt;stage4 THEN &#39;Stage II&#39;</span></a>
<a class="sourceLine" id="cb12-5" title="5"><span class="st">  WHEN stage3&lt;stage1 AND stage3&lt;stage2 AND stage3&lt;stage4 THEN &#39;Stage III&#39;</span></a>
<a class="sourceLine" id="cb12-6" title="6"><span class="st">  ELSE &#39;Stage IV&#39; END) AS prediction</span></a>
<a class="sourceLine" id="cb12-7" title="7"><span class="st">FROM distance)</span></a>
<a class="sourceLine" id="cb12-8" title="8"></a>
<a class="sourceLine" id="cb12-9" title="9"><span class="st">SELECT a.case_barcode, a.prediction, b.clinical_stage</span></a>
<a class="sourceLine" id="cb12-10" title="10"><span class="st">FROM pred AS a</span></a>
<a class="sourceLine" id="cb12-11" title="11"><span class="st">JOIN clin AS b</span></a>
<a class="sourceLine" id="cb12-12" title="12"><span class="st">ON a.case_barcode = b.case_barcode</span></a>
<a class="sourceLine" id="cb12-13" title="13"><span class="st">&quot;</span></a></code></pre></div>
<p>Now that we have our full query laid out, we can combine all of the sub-queries into the full string and query the data in BigQuery.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="co"># Build the query with all of the subqueries</span></a>
<a class="sourceLine" id="cb13-2" title="2">nc_query &lt;-<span class="st"> </span><span class="kw">str_c</span>(filtered_clin, filtered_expr, centroids, test, dist, pred)</a>
<a class="sourceLine" id="cb13-3" title="3"><span class="co"># To see the R console output with query processing information, turn queit to FALSE</span></a>
<a class="sourceLine" id="cb13-4" title="4">nc_result &lt;-<span class="st"> </span><span class="kw">bq_project_query</span>(billing, nc_query, <span class="dt">quiet =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb13-5" title="5"><span class="co"># Transform the query result into a tibble</span></a>
<a class="sourceLine" id="cb13-6" title="6">nc &lt;-<span class="st"> </span><span class="kw">bq_table_download</span>(nc_result, <span class="dt">quiet =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb13-7" title="7">nc</a></code></pre></div>
<pre><code>## # A tibble: 40 x 3
##    case_barcode prediction clinical_stage
##    &lt;chr&gt;        &lt;chr&gt;      &lt;chr&gt;         
##  1 TCGA-BQ-7055 Stage I    Stage I       
##  2 TCGA-2Z-A9JK Stage IV   Stage III     
##  3 TCGA-EV-5903 Stage IV   Stage I       
##  4 TCGA-2Z-A9JE Stage I    Stage I       
##  5 TCGA-G7-6797 Stage IV   Stage III     
##  6 TCGA-AL-3473 Stage IV   Stage II      
##  7 TCGA-A4-A4ZT Stage II   Stage II      
##  8 TCGA-BQ-5881 Stage I    Stage I       
##  9 TCGA-B1-5398 Stage II   Stage II      
## 10 TCGA-A4-A57E Stage I    Stage IV      
## # ... with 30 more rows
</code></pre>
<h2 id="find-accuracy-of-the-model">Find Accuracy of the Model</h2>
<p>To find the accuracy of the model, we weil use a confusion matrix and then will calculate the accuracy with the following formula<sup>3</sup>:</p>
<p>[Accuracy = \frac{M_{ii}}{\sum_{j}M_{ji}}]</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1"><span class="co"># Create a confusion matrix for the clinical predictions and</span></a>
<a class="sourceLine" id="cb15-2" title="2"><span class="co"># calculate the accuracy</span></a>
<a class="sourceLine" id="cb15-3" title="3"><span class="kw">confusionMatrix</span>(<span class="kw">as.factor</span>(nc<span class="op">$</span>prediction),<span class="kw">as.factor</span>(nc<span class="op">$</span>clinical_stage))</a></code></pre></div>
<pre><code>## Confusion Matrix and Statistics
## 
##            Reference
## Prediction  Stage I Stage II Stage III Stage IV
##   Stage I        12        0         1        2
##   Stage II        3        2         1        0
##   Stage III       1        0         0        0
##   Stage IV       13        2         3        0
## 
## Overall Statistics
##                                           
##                Accuracy : 0.35            
##                  95% CI : (0.2063, 0.5168)
##     No Information Rate : 0.725           
##     P-Value [Acc &gt; NIR] : 1.000000        
##                                           
##                   Kappa : 0.0545          
##                                           
##  Mcnemar&#39;s Test P-Value : 0.009041        
## 
## Statistics by Class:
## 
##                      Class: Stage I Class: Stage II Class: Stage III Class: Stage IV
## Sensitivity                  0.4138          0.5000           0.0000          0.0000
## Specificity                  0.7273          0.8889           0.9714          0.5263
## Pos Pred Value               0.8000          0.3333           0.0000          0.0000
## Neg Pred Value               0.3200          0.9412           0.8718          0.9091
## Prevalence                   0.7250          0.1000           0.1250          0.0500
## Detection Rate               0.3000          0.0500           0.0000          0.0000
## Detection Prevalence         0.3750          0.1500           0.0250          0.4500
## Balanced Accuracy            0.5705          0.6944           0.4857          0.2632
</code></pre>
<p>It seems that this model has an overall accuracy of 35% and was able to get close to 41% accuracy for Stage 1 but it was not able to accurately predict any of the other classes. We can tell that this is not a good model for this data set though it could easily be updated to different genes or data sets.</p>
<h1 id="references">References</h1>
<p><small><sup>1</sup>Statistical classification - Wikipedia. n.d. <a href="https://en.wikipedia.org/wiki/Statistical_classification">https://en.wikipedia.org/wiki/Statistical_classification</a>. </small></p>
<p><small><sup>2</sup>Euclidean distance - Wikipedia. n.d. <a href="https://en.wikipedia.org/wiki/Euclidean_distance">https://en.wikipedia.org/wiki/Euclidean_distance</a>.</small></p>
<p><small><sup>3</sup>Finding Nearest Neighbors in SQL | Sisense. n.d. <a href="https://www.sisense.com/blog/finding-nearest-neighbors-in-sql/">https://www.sisense.com/blog/finding-nearest-neighbors-in-sql/</a>.</small></p>
<p><small><sup>4</sup>Machine Learning with Python: Confusion Matrix in Machine Learning with Python. n.d. <a href="https://www.python-course.eu/confusion_matrix.php">https://www.python-course.eu/confusion_matrix.php</a>.</small></p>

</body>
</html>
